pysages/ssages/methods/umbrella.py

```diff
@@ -0,0 +1,49 @@
+# SPDX-License-Identifier: MIT
+# Copyright (c) 2020-2021: PySAGES contributors
+# See LICENSE.md and CONTRIBUTORS.md at https://github.com/SSAGESLabs/PySAGES
+
+from typing import NamedTuple
+from .core import SamplingMethod, generalize  # pylint: disable=relative-beyond-top-level
+import jax.numpy as np
+from jaxlib.xla_extension import DeviceArray as JaxArray
+
+
+class UmbrellaState(NamedTuple):
+    """
+    Description of an umbrella sampling state.
+
+    bias -- additional forces for the class.
+    xi -- current cv value.
+    """
+    bias: JaxArray
+    xi: JaxArray
+
+    def __repr__(self):
+        return repr("PySAGES" + type(self).__name__)
+
+
+class UmbrellaSampling(SamplingMethod):
+    def __init__(self, cvs, kspring, center, *args, **kwargs):
+        super().__init__(cvs, args, kwargs)
+        self.kspring = np.asarray(kspring)
+        self.center = np.asarray(center)
+
+    def __call__(self, snapshot, helpers):
+        return _umbrella(snapshot, self.cv, self.center, self.kspring, helpers)
+
+
+def _umbrella(snapshot, cv, center, kspring, helpers):
+    natoms = np.size(snapshot.positions, 0)
+    indices = helpers.indices
+
+    def initialize():
+        bias = np.zeros((natoms, 3))
+        return UmbrellaState(bias, None)
+
+    def update(state, rs, vms, ids):
+        xi, Jxi = cv(rs, indices(ids))
+        D = kspring * (xi - center)
+        bias = (-D * Jxi).reshape(state.bias.shape)
```

Comment:
```suggestion
        bias = (-Jxi.T @ D).reshape(state.bias.shape)
```
does this and passing the following

```python
cvs = (
    Component([0], 0),
    Component([0], 1),
)

center_cv = [[0., 0.]]
```
work?

Reply:
No, unfortunately it does not.
The shapes of the bias seems to mismatch if I try this.

```
  File "/home/ludwig/git/PySAGES/examples/umbrella_sampling/umbrella.py", line 57, in <module>
    main()
  File "/home/ludwig/git/PySAGES/examples/umbrella_sampling/umbrella.py", line 52, in main
    hoomd.run(1e5, limit_multiple=100)
  File "/home/ludwig/git/hoomd-blue/install/hoomd/__init__.py", line 201, in run
    context.current.system.run(int(tsteps), callback_period, callback, limit_hours, int(limit_multiple));
  File "/home/ludwig/.local/lib/python3.9/site-packages/pysages/backends/hoomd.py", line 44, in update
    self.state = self._update(self.snapshot, self.state)
  File "/home/ludwig/.local/lib/python3.9/site-packages/jax/_src/traceback_util.py", line 162, in reraise_with_filtered_traceback
    return fun(*args, **kwargs)
  File "/home/ludwig/.local/lib/python3.9/site-packages/jax/_src/api.py", line 411, in cache_miss
    out_flat = xla.xla_call(
  File "/home/ludwig/.local/lib/python3.9/site-packages/jax/core.py", line 1618, in bind
    return call_bind(self, fun, *args, **params)
  File "/home/ludwig/.local/lib/python3.9/site-packages/jax/core.py", line 1609, in call_bind
    outs = primitive.process(top_trace, fun, tracers, params)
  File "/home/ludwig/.local/lib/python3.9/site-packages/jax/core.py", line 1621, in process
    return trace.process_call(self, fun, tracers, params)
  File "/home/ludwig/.local/lib/python3.9/site-packages/jax/core.py", line 615, in process_call
    return primitive.impl(f, *tracers, **params)
  File "/home/ludwig/.local/lib/python3.9/site-packages/jax/interpreters/xla.py", line 622, in _xla_call_impl
    compiled_fun = _xla_callable(fun, device, backend, name, donated_invars,
  File "/home/ludwig/.local/lib/python3.9/site-packages/jax/linear_util.py", line 262, in memoized_fun
    ans = call(fun, *args)
  File "/home/ludwig/.local/lib/python3.9/site-packages/jax/interpreters/xla.py", line 694, in _xla_callable
    return lower_xla_callable(fun, device, backend, name, donated_invars, *arg_specs).compile().unsafe_call
  File "/home/ludwig/.local/lib/python3.9/site-packages/jax/interpreters/xla.py", line 702, in lower_xla_callable
    jaxpr, out_avals, consts = pe.trace_to_jaxpr_final(
  File "/home/ludwig/.local/lib/python3.9/site-packages/jax/interpreters/partial_eval.py", line 1522, in trace_to_jaxpr_final
    jaxpr, out_avals, consts = trace_to_subjaxpr_dynamic(fun, main, in_avals)
  File "/home/ludwig/.local/lib/python3.9/site-packages/jax/interpreters/partial_eval.py", line 1500, in trace_to_subjaxpr_dynamic
    ans = fun.call_wrapped(*in_tracers)
  File "/home/ludwig/.local/lib/python3.9/site-packages/jax/linear_util.py", line 166, in call_wrapped
    ans = self.f(*args, **dict(self.params, **kwargs))
  File "/home/ludwig/.local/lib/python3.9/site-packages/pysages/ssages/methods/core.py", line 74, in update
    return _update(state, rs, vms, ids)
  File "/home/ludwig/.local/lib/python3.9/site-packages/jax/_src/traceback_util.py", line 162, in reraise_with_filtered_traceback
    return fun(*args, **kwargs)
  File "/home/ludwig/.local/lib/python3.9/site-packages/jax/_src/api.py", line 411, in cache_miss
    out_flat = xla.xla_call(
  File "/home/ludwig/.local/lib/python3.9/site-packages/jax/core.py", line 1618, in bind
    return call_bind(self, fun, *args, **params)
  File "/home/ludwig/.local/lib/python3.9/site-packages/jax/core.py", line 1609, in call_bind
    outs = primitive.process(top_trace, fun, tracers, params)
  File "/home/ludwig/.local/lib/python3.9/site-packages/jax/core.py", line 1621, in process
    return trace.process_call(self, fun, tracers, params)
  File "/home/ludwig/.local/lib/python3.9/site-packages/jax/interpreters/partial_eval.py", line 1310, in process_call
    jaxpr, out_avals, consts = trace_to_subjaxpr_dynamic(f, self.main, in_avals)
  File "/home/ludwig/.local/lib/python3.9/site-packages/jax/interpreters/partial_eval.py", line 1500, in trace_to_subjaxpr_dynamic
    ans = fun.call_wrapped(*in_tracers)
  File "/home/ludwig/.local/lib/python3.9/site-packages/jax/linear_util.py", line 166, in call_wrapped
    ans = self.f(*args, **dict(self.params, **kwargs))
  File "/home/ludwig/.local/lib/python3.9/site-packages/pysages/ssages/methods/umbrella.py", line 46, in update
    bias = (-D * Jxi).reshape(state.bias.shape)
  File "/home/ludwig/.local/lib/python3.9/site-packages/jax/core.py", line 515, in __mul__
    def __mul__(self, other): return self.aval._mul(self, other)
  File "/home/ludwig/.local/lib/python3.9/site-packages/jax/_src/numpy/lax_numpy.py", line 6553, in deferring_binary_op
    return binary_op(self, other)
  File "/home/ludwig/.local/lib/python3.9/site-packages/jax/_src/traceback_util.py", line 162, in reraise_with_filtered_traceback
    return fun(*args, **kwargs)
  File "/home/ludwig/.local/lib/python3.9/site-packages/jax/_src/api.py", line 411, in cache_miss
    out_flat = xla.xla_call(
  File "/home/ludwig/.local/lib/python3.9/site-packages/jax/core.py", line 1618, in bind
    return call_bind(self, fun, *args, **params)
  File "/home/ludwig/.local/lib/python3.9/site-packages/jax/core.py", line 1609, in call_bind
    outs = primitive.process(top_trace, fun, tracers, params)
  File "/home/ludwig/.local/lib/python3.9/site-packages/jax/core.py", line 1621, in process
    return trace.process_call(self, fun, tracers, params)
  File "/home/ludwig/.local/lib/python3.9/site-packages/jax/interpreters/partial_eval.py", line 1310, in process_call
    jaxpr, out_avals, consts = trace_to_subjaxpr_dynamic(f, self.main, in_avals)
  File "/home/ludwig/.local/lib/python3.9/site-packages/jax/interpreters/partial_eval.py", line 1500, in trace_to_subjaxpr_dynamic
    ans = fun.call_wrapped(*in_tracers)
  File "/home/ludwig/.local/lib/python3.9/site-packages/jax/linear_util.py", line 166, in call_wrapped
    ans = self.f(*args, **dict(self.params, **kwargs))
  File "/home/ludwig/.local/lib/python3.9/site-packages/jax/_src/numpy/lax_numpy.py", line 686, in fn
    return lax_fn(x1, x2) if x1.dtype != bool_ else bool_lax_fn(x1, x2)
  File "/home/ludwig/.local/lib/python3.9/site-packages/jax/_src/lax/lax.py", line 354, in mul
    return mul_p.bind(x, y)
  File "/home/ludwig/.local/lib/python3.9/site-packages/jax/core.py", line 267, in bind
    out = top_trace.process_primitive(self, tracers, params)
  File "/home/ludwig/.local/lib/python3.9/site-packages/jax/interpreters/partial_eval.py", line 1297, in process_primitive
    out_avals = primitive.abstract_eval(*avals, **params)
  File "/home/ludwig/.local/lib/python3.9/site-packages/jax/_src/lax/lax.py", line 2151, in standard_abstract_eval
    return ShapedArray(shape_rule(*avals, **kwargs), dtype_rule(*avals, **kwargs),
  File "/home/ludwig/.local/lib/python3.9/site-packages/jax/_src/lax/lax.py", line 2247, in _broadcasting_shape_rule
    raise TypeError(msg.format(name, ', '.join(map(str, map(tuple, shapes)))))
jax._src.traceback_util.UnfilteredStackTrace: TypeError: mul got incompatible shapes for broadcasting: (1, 2), (2, 600).

The stack trace below excludes JAX-internal frames.
The preceding is the original exception that occurred, unmodified.

--------------------

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/home/ludwig/git/PySAGES/examples/umbrella_sampling/umbrella.py", line 57, in <module>
    main()
  File "/home/ludwig/git/PySAGES/examples/umbrella_sampling/umbrella.py", line 52, in main
    hoomd.run(1e5, limit_multiple=100)
  File "/home/ludwig/git/hoomd-blue/install/hoomd/__init__.py", line 201, in run
    context.current.system.run(int(tsteps), callback_period, callback, limit_hours, int(limit_multiple));
  File "/home/ludwig/.local/lib/python3.9/site-packages/pysages/backends/hoomd.py", line 44, in update
    self.state = self._update(self.snapshot, self.state)
  File "/home/ludwig/.local/lib/python3.9/site-packages/pysages/ssages/methods/core.py", line 74, in update
    return _update(state, rs, vms, ids)
  File "/home/ludwig/.local/lib/python3.9/site-packages/pysages/ssages/methods/umbrella.py", line 46, in update
    bias = (-D * Jxi).reshape(state.bias.shape)
  File "/home/ludwig/.local/lib/python3.9/site-packages/jax/_src/numpy/lax_numpy.py", line 6553, in deferring_binary_op
    return binary_op(self, other)
  File "/home/ludwig/.local/lib/python3.9/site-packages/jax/_src/numpy/lax_numpy.py", line 686, in fn
    return lax_fn(x1, x2) if x1.dtype != bool_ else bool_lax_fn(x1, x2)
TypeError: mul got incompatible shapes for broadcasting: (1, 2), (2, 600).

Reply:
I tried debugging it a little bit by hand, but couldn't find a handle of how to deal with it.

Reply:
You're stacktrace is showing `bias = (-D * Jxi).reshape(state.bias.shape)`. Did you checked with: `bias = (-Jxi.T @ D).reshape(state.bias.shape)`

Reply:
it works for the one 1D case, but not with more CVs
but I think you are on the right path

Reply:
Do you have a stacktrace for the 2D case?

Reply:
```
Traceback (most recent call last):
  File "umbrella.py", line 51, in <module>
    main()
  File "umbrella.py", line 46, in main
    hoomd.run(1e5, limit_multiple=100)
  File "/software/python-anaconda-2020.11-el8-x86_64/envs/pysages/lib/python3.8/site-packages/hoomd/__init__.py", line 201, in run
    context.current.system.run(int(tsteps), callback_period, callback, limit_hours, int(limit_multiple));
  File "/home/ludwigschneider/.local/lib/python3.8/site-packages/pysages/backends/hoomd.py", line 44, in update
    self.state = self._update(self.snapshot, self.state)
  File "/home/ludwigschneider/.local/lib/python3.8/site-packages/jax/_src/traceback_util.py", line 183, in reraise_with_filtered_traceback
    return fun(*args, **kwargs)
  File "/home/ludwigschneider/.local/lib/python3.8/site-packages/jax/_src/api.py", line 399, in cache_miss
    out_flat = xla.xla_call(
  File "/home/ludwigschneider/.local/lib/python3.8/site-packages/jax/core.py", line 1561, in bind
    return call_bind(self, fun, *args, **params)
  File "/home/ludwigschneider/.local/lib/python3.8/site-packages/jax/core.py", line 1552, in call_bind
    outs = primitive.process(top_trace, fun, tracers, params)
  File "/home/ludwigschneider/.local/lib/python3.8/site-packages/jax/core.py", line 1564, in process
    return trace.process_call(self, fun, tracers, params)
  File "/home/ludwigschneider/.local/lib/python3.8/site-packages/jax/core.py", line 607, in process_call
    return primitive.impl(f, *tracers, **params)
  File "/home/ludwigschneider/.local/lib/python3.8/site-packages/jax/interpreters/xla.py", line 607, in _xla_call_impl
    compiled_fun = _xla_callable(fun, device, backend, name, donated_invars,
  File "/home/ludwigschneider/.local/lib/python3.8/site-packages/jax/linear_util.py", line 262, in memoized_fun
    ans = call(fun, *args)
  File "/home/ludwigschneider/.local/lib/python3.8/site-packages/jax/interpreters/xla.py", line 682, in _xla_callable
    jaxpr, out_avals, consts = pe.trace_to_jaxpr_final(
  File "/home/ludwigschneider/.local/lib/python3.8/site-packages/jax/interpreters/partial_eval.py", line 1285, in trace_to_jaxpr_final
    jaxpr, out_avals, consts = trace_to_subjaxpr_dynamic(fun, main, in_avals)
  File "/home/ludwigschneider/.local/lib/python3.8/site-packages/jax/interpreters/partial_eval.py", line 1263, in trace_to_subjaxpr_dynamic
    ans = fun.call_wrapped(*in_tracers)
  File "/home/ludwigschneider/.local/lib/python3.8/site-packages/jax/linear_util.py", line 166, in call_wrapped
    ans = self.f(*args, **dict(self.params, **kwargs))
  File "/home/ludwigschneider/.local/lib/python3.8/site-packages/pysages/ssages/methods/core.py", line 74, in update
    return _update(state, rs, vms, ids)
  File "/home/ludwigschneider/.local/lib/python3.8/site-packages/jax/_src/traceback_util.py", line 183, in reraise_with_filtered_traceback
    return fun(*args, **kwargs)
  File "/home/ludwigschneider/.local/lib/python3.8/site-packages/jax/_src/api.py", line 399, in cache_miss
    out_flat = xla.xla_call(
  File "/home/ludwigschneider/.local/lib/python3.8/site-packages/jax/core.py", line 1561, in bind
    return call_bind(self, fun, *args, **params)
  File "/home/ludwigschneider/.local/lib/python3.8/site-packages/jax/core.py", line 1552, in call_bind
    outs = primitive.process(top_trace, fun, tracers, params)
  File "/home/ludwigschneider/.local/lib/python3.8/site-packages/jax/core.py", line 1564, in process
    return trace.process_call(self, fun, tracers, params)
  File "/home/ludwigschneider/.local/lib/python3.8/site-packages/jax/interpreters/partial_eval.py", line 1072, in process_call
    jaxpr, out_avals, consts = trace_to_subjaxpr_dynamic(f, self.main, in_avals)
  File "/home/ludwigschneider/.local/lib/python3.8/site-packages/jax/interpreters/partial_eval.py", line 1263, in trace_to_subjaxpr_dynamic
    ans = fun.call_wrapped(*in_tracers)
  File "/home/ludwigschneider/.local/lib/python3.8/site-packages/jax/linear_util.py", line 166, in call_wrapped
    ans = self.f(*args, **dict(self.params, **kwargs))
  File "/home/ludwigschneider/.local/lib/python3.8/site-packages/pysages/ssages/methods/umbrella.py", line 46, in update
    bias = (-D * Jxi).reshape(state.bias.shape)
  File "/home/ludwigschneider/.local/lib/python3.8/site-packages/jax/core.py", line 507, in __mul__
    def __mul__(self, other): return self.aval._mul(self, other)
  File "/home/ludwigschneider/.local/lib/python3.8/site-packages/jax/_src/numpy/lax_numpy.py", line 5871, in deferring_binary_op
    return binary_op(self, other)
  File "/home/ludwigschneider/.local/lib/python3.8/site-packages/jax/_src/numpy/lax_numpy.py", line 431, in fn
    return lax_fn(x1, x2) if x1.dtype != bool_ else bool_lax_fn(x1, x2)
  File "/home/ludwigschneider/.local/lib/python3.8/site-packages/jax/_src/lax/lax.py", line 348, in mul
    return mul_p.bind(x, y)
  File "/home/ludwigschneider/.local/lib/python3.8/site-packages/jax/core.py", line 264, in bind
    out = top_trace.process_primitive(self, tracers, params)
  File "/home/ludwigschneider/.local/lib/python3.8/site-packages/jax/interpreters/partial_eval.py", line 1059, in process_primitive
    out_avals = primitive.abstract_eval(*avals, **params)
  File "/home/ludwigschneider/.local/lib/python3.8/site-packages/jax/_src/lax/lax.py", line 2136, in standard_abstract_eval
    return ShapedArray(shape_rule(*avals, **kwargs), dtype_rule(*avals, **kwargs),
  File "/home/ludwigschneider/.local/lib/python3.8/site-packages/jax/_src/lax/lax.py", line 2232, in _broadcasting_shape_rule
    raise TypeError(msg.format(name, ', '.join(map(str, map(tuple, shapes)))))
jax._src.traceback_util.UnfilteredStackTrace: TypeError: mul got incompatible shapes for broadcasting: (1, 3), (3, 600).

The stack trace below excludes JAX-internal frames.
The preceding is the original exception that occurred, unmodified.

--------------------

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "umbrella.py", line 51, in <module>
    main()
  File "umbrella.py", line 46, in main
    hoomd.run(1e5, limit_multiple=100)
  File "/software/python-anaconda-2020.11-el8-x86_64/envs/pysages/lib/python3.8/site-packages/hoomd/__init__.py", line 201, in run
    context.current.system.run(int(tsteps), callback_period, callback, limit_hours, int(limit_multiple));
  File "/home/ludwigschneider/.local/lib/python3.8/site-packages/pysages/backends/hoomd.py", line 44, in update
    self.state = self._update(self.snapshot, self.state)
  File "/home/ludwigschneider/.local/lib/python3.8/site-packages/pysages/ssages/methods/core.py", line 74, in update
    return _update(state, rs, vms, ids)
  File "/home/ludwigschneider/.local/lib/python3.8/site-packages/pysages/ssages/methods/umbrella.py", line 46, in update
    bias = (-D * Jxi).reshape(state.bias.shape)
  File "/home/ludwigschneider/.local/lib/python3.8/site-packages/jax/_src/numpy/lax_numpy.py", line 5871, in deferring_binary_op
    return binary_op(self, other)
  File "/home/ludwigschneider/.local/lib/python3.8/site-packages/jax/_src/numpy/lax_numpy.py", line 431, in fn
    return lax_fn(x1, x2) if x1.dtype != bool_ else bool_lax_fn(x1, x2)
TypeError: mul got incompatible shapes for broadcasting: (1, 3), (3, 600).
[midway3-login1.rcc.local:mpi_rank_0][deallocate_vbuf_region] [src/mpid/ch3/channels/mrail/src/gen2/vbuf.c:238] cuda failed with 3 (initialization error)